radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/core/handlers/OlympiaEndStateHandler.java

```diff
@@ -62,70 +62,97 @@
   private final EngineStore<LedgerAndBFTProof> engineStore;
   private final CurrentForkView currentForkView;
+  private final Hasher hasher;
+  private final HashSigner hashSigner;
 
-  private Optional<byte[]> cachedCompressedEndStateBytes = Optional.empty();
+  private Optional<OlympiaEndStateResponse> cachedEndStateResponse = Optional.empty();
 
   @Inject
   OlympiaEndStateHandler(
       RadixEngine<LedgerAndBFTProof> radixEngine,
       EngineStore<LedgerAndBFTProof> engineStore,
-      CurrentForkView currentForkView) {
+      CurrentForkView currentForkView,
+      Hasher hasher,
+      HashSigner hashSigner) {
+    super(OlympiaEndStateRequest.class);
     this.radixEngine = Objects.requireNonNull(radixEngine);
     this.engineStore = Objects.requireNonNull(engineStore);
-    this.currentForkView = currentForkView;
+    this.currentForkView = Objects.requireNonNull(currentForkView);
+    this.hasher = Objects.requireNonNull(hasher);
+    this.hashSigner = Objects.requireNonNull(hashSigner);
   }
 
   @Override
-  public void handleRequest(HttpServerExchange exchange) throws Exception {
+  public OlympiaEndStateResponse handleRequest(OlympiaEndStateRequest request)
+      throws CoreApiException {
     if (!radixEngine.isShutDown()) {
-      // The engine is still running, can't create the end state
-      exchange.setStatusCode(409);
-      exchange.getResponseSender().send("The engine is still running");
+      // Radix Engine is still running
+      throw CoreApiException.badRequest(
+          new EngineIsNotShutDownError()
+              .message("Can't create the end state. Engine isn't yet shut down.")
+              .type(EngineIsNotShutDownError.class.getSimpleName()));
     }
 
-    // THe engine is shut down. Good. We can safely create the end state and cache it
-    final var compressedEndStateBytes = getEndStateFromCacheOrPrepareIfNeeded();
-    final var result = Bytes.toHexString(compressedEndStateBytes);
-
-    exchange.setStatusCode(200);
-    exchange.getResponseSender().send(result);
+    synchronized (endStateLock) {
+      try {
+        if (this.cachedEndStateResponse.isEmpty()) {
+          prepareEndStateResponseAndSaveToCache();
+        }
+        return this.cachedEndStateResponse.orElseThrow();
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    }
   }
 
-  private byte[] getEndStateFromCacheOrPrepareIfNeeded() throws IOException {
-    if (this.cachedCompressedEndStateBytes.isPresent()) {
-      return this.cachedCompressedEndStateBytes.orElseThrow();
-    }
+  private void prepareEndStateResponseAndSaveToCache() throws IOException {
+    final var endStateBytes = prepareEndState();
+    final var hash = hasher.hashBytes(endStateBytes);
+    final var signature = hashSigner.sign(hash);
+    final var response =
+        new OlympiaEndStateResponse()
+            .hash(Bytes.toHexString(hash.asBytes()))
+            .signature(signature.toHexString())
```

Comment:
The `toHexString` method of signature is (I think) un-invertible, and flawed.

We should really use DER encoding, but I don't think we have any method in the production code to create a DER encoded sig (only to decode)... However, we do have it in `ConstructionFinalizeTest`.

So, copying from there... could we create the following method on `ECDSASignature`, and actually use that to return a hex-encoded DER signature here?

```
public byte[] toUnrecoverableDERBytes() {
    var os = new ByteArrayOutputStream();
    var asn1OutputStream = ASN1OutputStream.create(os);
    asn1OutputStream.writeObject(
        new DLSequence(
            new ASN1Encodable[] {new ASN1Integer(sig.getR()), new ASN1Integer(sig.getS())}));
    return os.toByteArray();
}
```

Can we also:
* Update ConstructionFinalizeTest to use this; to check we've done it correctly
* Replace `toHexString()` to be `Bytes.toHexString(toUnrecoverableDERBytes())`

Reply:
done